### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Eli-Ionescu**
Iteration 1 

The core functionalities for the Distraction Shield extension.

### CODE REVIEW

File: sampleExtension/background.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+initExtension = function () {
+    //First receive the blacklist from the sync storage, and then create a onBeforeRequest listener using this list.
+    updateBlockedSites(replaceListener);
+    addBrowserActionListener();
+};
+
+// This function receives the blacklist from the sync storage.
+updateBlockedSites = function(callback){
+    chrome.storage.sync.get("tds_blacklist", function (items) {
+        if (!chrome.runtime.error) {
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: Why isn't this runtime error logged?

File: sampleExtension/background.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            blockedSites = items.tds_blacklist;
+            return callback();
+        }
+    });
+};
+
+// This function adds one url to the blacklist
+addToBlockedSites = function(urlToAdd) {
+    var formatUrl = urlToAdd.split(['//'])[1];
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: Different levels of abstraction in one function, introduce auxilary function, e.g. `formatTheUrl()`.

File: sampleExtension/background.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            },
+            ["blocking"]
+        );
+    }
+};
+
+intercept = function() {
+    incrementInterceptionCounter();
+    return {redirectUrl: "https://zeeguu.herokuapp.com/getex"};
+    //return {redirectUrl: "https://zeeguu.herokuapp.com/getex"};
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: Why the same line in the comment here.

File: sampleExtension/background.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    types: ["main_frame"]
+            },
+            ["blocking"]
+        );
+    }
+};
+
+intercept = function() {
+    incrementInterceptionCounter();
+    return {redirectUrl: "https://zeeguu.herokuapp.com/getex"};
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: What a pretty magical string, may be you should collect all the magical strings in their own file. I.e. no magic constants.

File: sampleExtension/background.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+
+/* --------------- ---- Run upon installation ---- ---------------*/
+
+chrome.runtime.onInstalled.addListener(function() {
+    chrome.storage.sync.get("tds_blacklist", function(output) {
+        if (output.tds_blacklist == null) {
+            chrome.storage.sync.set({"tds_blacklist": []}, function () {
+                if (chrome.runtime.error) {
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: Given the frequency with which you handle chrome runtime errors, wouldn't make sense to make a checkForRunTimeError() or something like it. If you ever decide to handle the runtime error not by printing to the console but by doing something else you only have to implement that in one place.

File: sampleExtension/background.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                    console.log("Runtime error.");
+                }
+            });
+        }
+    });
+    chrome.storage.sync.get("tds_interceptCounter", function(output) {
+        if (output.tds_interceptCounter == null) {
```
Comment (lauraBaakman) [inline sampleExtension/background.js]: Too many levels of abstraction in one function.

File: sampleExtension/optionsPage/options.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
+
+// Log console messages to the background page console instead of the content page.
+var console = chrome.extension.getBackgroundPage().console;
+
+var html_blacklist = $('#blacklistedSites');
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: 👍

File: sampleExtension/optionsPage/options.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+var html_intCnt = $('#iCounter');
+//Local variable that holds the list of links that we are blacklisting.
+var links = [];
+var interceptionCounter = 0;
+
+//Initialize HTML elements and set the local variables
+setLinksAndBlacklistList = function() {
+    chrome.storage.sync.get("tds_blacklist", function(output) {
+        links = output.tds_blacklist;
+        if (!chrome.runtime.error) {
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: What if you have a runtime error?

File: sampleExtension/optionsPage/options.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+var html_blacklist = $('#blacklistedSites');
+var html_txtFld = $('#textFld');
+var html_intCnt = $('#iCounter');
+//Local variable that holds the list of links that we are blacklisting.
+var links = [];
+var interceptionCounter = 0;
+
+//Initialize HTML elements and set the local variables
+setLinksAndBlacklistList = function() {
+    chrome.storage.sync.get("tds_blacklist", function(output) {
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: Too many levels of abstraction in one function.

File: sampleExtension/optionsPage/options.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        if (!chrome.runtime.error) {
+            //For every element in the array append it to the html blacklist
+            $.each(links, function(key, value) {
+                html_blacklist.append($("<option></option>").text(value));
+            });
+        }
+    });
+    chrome.storage.sync.get("tds_interceptCounter", function(output) {
+        interceptionCounter = output.tds_interceptCounter;
+        if (!chrome.runtime.error) {
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: What if you have an error?

File: sampleExtension/optionsPage/options.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+saveButtonClick = function() {
+    var newurl = html_txtFld.val();
+    // Add the url to the sync storage.
+    links.push("*://"+newurl+"/*");
+    chrome.storage.sync.set({"tds_blacklist" : links }, function() {
+        if(chrome.runtime.error) {
+            console.log("Runtime error.");
+        }
+        //Append element to the html blacklist
+        html_blacklist.append($("<option></option>").text("*://"+newurl+"/*"));
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: Too many levels of abstraction, the comment lines give a pretty good indiction of what the functions should be.

File: sampleExtension/optionsPage/options.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        // Empty the input box.
+        html_txtFld.val('');
+        var bg = chrome.extension.getBackgroundPage();
+        bg.updateBlockedSites(bg.replaceListener);
+    });
+};
+
+deleteButtonClick = function() {
+    var urltodelete = $("#blacklistedSites").find('option:selected');
```
Comment (lauraBaakman) [inline sampleExtension/optionsPage/options.js]: Too many levels of abstraction.

**Comment by DanChirtoaca** [comment]
For this iteration, the code you have written looks decent. 
Do take into consideration Laura's comments, especially the ones regarding hard coded constants and the extraction of specific levels of abstraction. Also, I would suggest looking into making your application components follow OOP paradigm, as you will extend it, it will help you a lot.

**Pull request merged**